src/adapter/src/coord/sequencer/inner/cluster.rs

```diff
@@ -428,4 +469,105 @@ impl Coordinator {
+                    Some(r.replica_id.clone())
+                } else {
+                    None
+                }
+            })
+            .collect_vec();
+        // Check For timeout
+        if Instant::now() > timeout_time {
+            // Timed out handle timeout action
+            match on_timeout {
+                OnTimeoutAction::Rollback => {
+                    self.active_conns
+                        .get_mut(session.conn_id())
+                        .expect("There must be an active connection")
+                        .pending_cluster_alters
+                        .remove(&cluster.id);
+                    let pending_replica_drops: Vec<DropObjectInfo> = pending_replicas
+                        .iter()
+                        .map(|r| {
+                            DropObjectInfo::ClusterReplica((
+                                cluster.id.clone(),
+                                r.to_owned(),
+                                ReplicaCreateDropReason::Manual,
+                            ))
+                        })
+                        .collect::<Vec<DropObjectInfo>>();
+                    self.catalog_transact(None, vec![Op::DropObjects(pending_replica_drops)])
+                        .await?;
+                    return Err(AdapterError::AlterClusterTimeout);
+                }
+                OnTimeoutAction::Commit => {
+                    let span = Span::current();
+                    return Ok(StageResult::Handle(mz_ore::task::spawn(
+                        || "Finalize Alter Cluster",
+                        async move {
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            let stage = ClusterStage::Finalize(AlterClusterFinalize {
+                                validity,
+                                plan,
+                                new_config,
+                            });
+                            Ok(Box::new(stage))
+                        }
+                        .instrument(span),
+                    )));
+                }
+            }
+        }
+        let hydrated = match self
+            .controller
+            .compute
+            .cluster_replicas_hydrated(cluster.id, pending_replicas)
+        {
+            Err(e) => coord_bail!("cluster not found {e}"),
+            Ok(b) => b,
+        };
+        if hydrated {
+            // We're done
+            Ok(StageResult::Immediate(Box::new(ClusterStage::Finalize(
+                AlterClusterFinalize {
+                    validity,
+                    plan,
+                    new_config,
+                },
+            ))))
+        } else {
+            // Check later
+            let span = Span::current();
+            Ok(StageResult::Handle(mz_ore::task::spawn(
+                || "Finalize Alter Cluster",
+                async move {
+                    tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
What's the 1 second sleep for?

Reply:
I'm not positive on this approach, the gist is that wait_for_pending_replicas_hydrated just re-enqueus the stage until it times out or the pending replicas have hydrated all flows.

This is somewhat modeled after the CheckSchedulingPolicies message which is kicked off ever ~3 seconds. I should at least make this a session var similar to CLUSTER_CHECK_ SCHEDULING_POLICY)INTERVAL.

Reply:
Oh OK, I see. Yeah making this a system configuration parameter would probably be useful.